turbine/src/cluster_nodes.rs

```diff
@@ -393,44 +407,90 @@ fn cmp_nodes_stake(a: &Node, b: &Node) -> Ordering {
+            match addrs.entry(addr) {
+                Entry::Occupied(mut entry) => {
+                    let conflict_idx: usize = *entry.get();
+                    let conflict_outset: u64 = nodes[conflict_idx].outset().unwrap_or(0);
+                    // Compute the index of the node that would get its TVU address removed
+                    let index_to_update = if conflict_outset > outset {
+                        // other entry is newer then whatever idx is pointing to
+                        // so we need to remove nodes[idx] TVU address
+                        idx
+                    } else {
+                        // set current index as "authoritative" owner of the address
+                        *entry.get_mut() = idx;
+                        //remove nodes[conflict_idx] TVU address
+                        conflict_idx
+                    };
+                    // this should never happen, but if it does we want to know
+                    if nodes[index_to_update].contact_info().unwrap().pubkey == keep_identity {
+                        warn!("Trying to remove my own TVU address!!!");
+                        debug_assert!(false, "Tried removing own TVU address");
+                    } else {
+                        nodes[index_to_update]
+                            .contact_info_mut()
+                            .unwrap()
+                            .remove_tvu_addr(protocol);
+                    }
+                }
+                Entry::Vacant(entry) => {
+                    entry.insert(idx);
+                }
+            }
             let count: usize = *counts
-                .entry((protocol, addr.ip()))
+                .entry(addr.ip())
                 .and_modify(|count| *count += 1)
                 .or_insert(1);
-            if !addrs.insert((protocol, addr)) || count > MAX_NUM_NODES_PER_IP_ADDRESS {
+            if count > MAX_NUM_NODES_PER_IP_ADDRESS {
                 // Remove the respective TVU address so that no more shreds are
                 // sent to this socket address.
-                node.remove_tvu_addr(protocol);
+                nodes[idx]
+                    .contact_info_mut()
+                    .unwrap()
+                    .remove_tvu_addr(protocol);
             }
         }
-        // Always keep staked nodes for deterministic shuffle,
-        // but drop non-staked nodes if they have no valid TVU address.
-        node_stake > 0u64
-            || TVU_PROTOCOLS
-                .into_iter()
-                .any(|protocol| node.tvu(protocol).is_some())
-    })
+    }
+    nodes.retain_mut(|node| {
+        let node_stake = node.stake;
+        // Always keep staked nodes for deterministic shuffle.
+        if node_stake > 0 {
+            return true;
+        }
+        // Do not purge the provided keep_identity to maintain datastructure invariants
+        if *node.pubkey() == keep_identity {
```

Comment:
the tree should look the same across all nodes in the cluster regardless of the node executing the code.
such kind of logic breaks that.

Reply:
It does not break determinism any more then the original implementation, i.e. it never wipes staked identities and sometimes wipes unstaked ones. If we want this to be 100% identical across unstaked nodes, we need to ensure gossip propagation of contactinfos is 100% identical too, which is clearly not poasible.